### PR TITLE
Protobuf generation for k8s.io/metrics

### PR DESCRIPTION
**Pull request by DirectXMan12**
This PR introduces protobuf generation for k8s.io/metrics.  Doing so required:

- fixing a bug in `go-to-protobuf` causing the `cast{key,value,type}` values to not be quoted when coming from struct tags (and not auto-injection by `go-to-protobuf` itself).
- Making sure the proto IDL in k8s.io/client-go had a package name of `k8s.io.client_go.xyz` and not `k8s.io.kubernetes.xyz`.

Additionally, I updated `go-to-protobuf` to skip functions and non-public types when composing the import list, which cuts down on the more bizarre imports in the IDL (like importing the sample API package in every IDL file because it contained `addToScheme`, like every other API package).

We use `castvalue` to force gogo-proto to realize that it should consider the value of the map which underlies `ResourceList` when calculating which imports need to be named.  Otherwise, it ignores the value's type, leading to compilation errors when it later can't find an import it assumed existed.  We accidentally didn't hit this in `k8s.io/kubernetes/pkg/api/v1` since another field coincidentally happens to directly use `resource.Quantity` (the value type of `ResourceList`).

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @smarterclayton hopefully I didn't break anything.  Does this seem reasonable?

@timstclair all you need from this should be the "rewriting package names" part (plus adding `-k8s.io/client-go/pkg/api/v1` to the list of packages in `go-to-protobuf`).

**Comment by DirectXMan12** [comment]
@caesarxuchao we forgot this earlier ;-)

File: cmd/libs/go2idl/go-to-protobuf/protobuf/cmd.go [hunk 1]
```diff
@@ -89,6 +89,9 @@ func New() *Generator {
 			`k8s.io/kubernetes/pkg/apis/settings/v1alpha1`,
 			`k8s.io/kubernetes/pkg/apis/storage/v1beta1`,
 			`k8s.io/kubernetes/pkg/apis/storage/v1`,
+			`-k8s.io/client-go/pkg/api/v1`,
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/cmd.go]: is "-" a typo?

File: staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go [hunk 2]
```diff
@@ -29,26 +29,26 @@ import (
...
 	// The following fields define time interval from which metrics were
 	// collected from the interval [Timestamp-Window, Timestamp].
-	Timestamp metav1.Time     `json:"timestamp"`
-	Window    metav1.Duration `json:"window"`
+	Timestamp metav1.Time     `json:"timestamp" protobuf:"bytes,2,opt,name=timestamp"`
+	Window    metav1.Duration `json:"window" protobuf:"bytes,3,opt,name=window"`
 
 	// The memory usage is the memory working set.
-	Usage v1.ResourceList `json:"usage"`
+	Usage v1.ResourceList `json:"usage" protobuf:"bytes,4,rep,name=usage,casttype=k8s.io/client-go/pkg/api/v1.ResourceList,castkey=k8s.io/client-go/pkg/api/v1.ResourceName,castvalue=k8s.io/apimachinery/pkg/api/resource.Quantity"`
```
Comment (caesarxuchao) [inline staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go]: ResourceList will be moved k8s.io/api in #44784, shall i just change the package name in casttype and castkey in #44784?

File: cmd/libs/go2idl/go-to-protobuf/protobuf/cmd.go (hunk 1 above)
Comment (DirectXMan12) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/cmd.go]: No, `-` means "consider types from this package, but don't generate IDL and Go code"

File: staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go (hunk 2 above)
Comment (DirectXMan12) [inline staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go]: Yeah.  You'll also probably need to change the list of files in the protobuf generator as well.

File: cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go [hunk 3]
```diff
@@ -167,6 +168,13 @@ func (n *protobufNamer) AssignTypesToPackages(c *generator.Context) error {
 			if t.Name.Package != p.PackagePath {
 				continue
 			}
+			// skip functions -- we don't care about them for protobuf
```
Comment (smarterclayton) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go]: these filters should be in assignGoTypeToProtoPackage

File: cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go [hunk 4]
```diff
@@ -167,6 +168,13 @@ func (n *protobufNamer) AssignTypesToPackages(c *generator.Context) error {
...
+			if t.Kind == types.Func || (t.Kind == types.DeclarationOf && t.Underlying.Kind == types.Func) {
+				continue
+			}
+			if unicode.IsLower([]rune(t.Name.Name)[0]) {
```
Comment (smarterclayton) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go]: we already have namer.IsPrivateGoName for this inside of the assigner

File: cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go (hunk 3 above)
Comment (smarterclayton) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go]: actually, i guess you can leave them here, but there's a more fundamental filter that seems like it is missing..

**Comment by smarterclayton** [comment]
A few nits, otherwise looks good

File: cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go (hunk 4 above)
Comment (DirectXMan12) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go]: ack, I'll use that

File: cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go (hunk 3 above)
Comment (DirectXMan12) [inline cmd/libs/go2idl/go-to-protobuf/protobuf/namer.go]: I've at least moved it out into its own function now.

**Comment by DirectXMan12** [comment]
ok, nits should be fixed

**Comment by smarterclayton** [comment]
The proto changes LGTM

**Comment by sttts** [comment]
Looks good. Rebase necessary.

**Comment by DirectXMan12** [comment]
@sttts rebased

**Comment by caesarxuchao** [comment]
@k8s-bot pull-kubernetes-unit test this@k8s-bot pull-kubernetes-unit test this

**Comment by caesarxuchao** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by caesarxuchao** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by smarterclayton** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by smarterclayton** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Pull request merged**